noir-projects/aztec-nr/aztec/src/state_vars/shared_mutable/shared_mutable.nr

```diff
@@ -0,0 +1,197 @@
+// measured in blocks. Reads in private are only valid as long as they are included in a block not too far into the 
+// future, so that they can guarantee the value will not have possibly changed by then (because of the delay).
+impl<T, DELAY> SharedMutable<T, DELAY> {
+    pub fn new(context: Context, storage_slot: Field) -> Self {
+        assert(storage_slot != 0, "Storage slot 0 not allowed. Storage slots must start from 1.");
+        Self { context, storage_slot }
+    }
+
+    pub fn schedule_value_change(self, new_value: T) {
+        let context = self.context.public.unwrap();
+        let mut scheduled_value_change: ScheduledValueChange<T, DELAY> = public_storage::read(self.get_derived_storage_slot());
+
+        scheduled_value_change.schedule_change(new_value, context.block_number() as u32);
+
+        public_storage::write(self.get_derived_storage_slot(), scheduled_value_change);
+    }
+
+    pub fn get_current_value_in_public(self) -> T {
+        let scheduled_value_change: ScheduledValueChange<T, DELAY> = public_storage::read(self.get_derived_storage_slot());
+
+        let block_number = self.context.public.unwrap().block_number() as u32;
+        scheduled_value_change.get_current_at(block_number)
+    }
+
+    pub fn get_scheduled_value_in_public(self) -> (T, u32) {
+        let scheduled_value_change: ScheduledValueChange<T, DELAY> = public_storage::read(self.get_derived_storage_slot());
+        scheduled_value_change.get_scheduled()
+    }
+
+    pub fn get_current_value_in_private(self) -> T where T: FromField {
+        let mut context = self.context.private.unwrap();
+
+        let (scheduled_value_change, historical_block_number) = self.historical_read_from_public_storage(*context);
+        let block_horizon = scheduled_value_change.get_block_horizon(historical_block_number);
+
+        context.request_max_block_number(block_horizon);
+        scheduled_value_change.get_current_at(historical_block_number)
+    }
+
+    fn historical_read_from_public_storage(
+        self,
+        context: PrivateContext
+    ) -> (ScheduledValueChange<T, DELAY>, u32) where T: FromField {
+        let derived_slot = self.get_derived_storage_slot();
+
+        // Ideally the following would be simply public_storage::read_historical, but we can't implement that yet.
+        let mut raw_fields = [0; 3];
+        for i in 0..3 {
+            raw_fields[i] = public_storage_historical_read(
+                    context,
+                    derived_slot + i as Field,
+                    context.this_address()
+                );
+        }
+
+        let scheduled_value: ScheduledValueChange<T, DELAY> = ScheduledValueChange::deserialize(raw_fields);
+        let historical_block_number = context.historical_header.global_variables.block_number as u32;
+
+        (scheduled_value, historical_block_number)
+    }
+
+    fn get_derived_storage_slot(self) -> Field {
```

Comment:
🫡  nice, already had jotted down that the macros should probably have a look to not have collisions